src/features/completions.zig

```diff
@@ -1495,11 +1007,654 @@ pub fn completionAtIndex(server: *Server, analyser: *Analyser, arena: std.mem.Al
+                braces_depth -= 1;
+                if (braces_depth < one_opening) return null;
+            },
+            .r_paren => {
+                if (braces_depth == one_opening and token_tags[token_index + 1] == .l_brace) break :find_r_paren_l_brace;
+            },
+            else => {},
+        }
+    }
+    // need at least `const name = switch(some_enum)` => 6 tokens min
+    if (token_index < 5) return null else token_index -= 1; // eat the `)`
+    const switch_condition_end_index = token_index;
+    // The following logic is just to make sure this is a switch and not a `for`
+    var parens_depth: i32 = even;
+    find_switch: while (token_index > 0) : (token_index -= 1) {
+        switch (token_tags[token_index]) {
+            .r_paren => parens_depth += 1,
+            .l_paren => {
+                parens_depth -= 1;
+                if (parens_depth == one_opening and token_tags[token_index - 1] == .keyword_switch) break :find_switch;
+            },
+            .semicolon => return null,
+            else => {},
+        }
+    }
+    if (token_index == 0) return null;
+
+    // const nts2 = std.time.nanoTimestamp();
+    // std.log.info("found switch in {} ns, tokens iterated: {}", .{ nts2 - nts1, starting_token_index - token_index });
+
+    return EnumLiteralContext{
+        .likely = .enum_literal,
+        .identifier_token_index = switch_condition_end_index,
+    };
+}
+
+/// Given a TypeWithHandle that is a container, adds it's `.container_field*`s to completions
+pub fn collectContainerFields(
+    arena: std.mem.Allocator,
+    container: Analyser.TypeWithHandle,
+    completions: *std.ArrayListUnmanaged(types.CompletionItem),
+) error{OutOfMemory}!void {
+    const node = switch (container.type.data) {
+        .other => |n| n,
+        else => return,
+    };
+    var buffer: [2]Ast.Node.Index = undefined;
+    const container_decl = Ast.fullContainerDecl(container.handle.tree, &buffer, node) orelse return;
+    for (container_decl.ast.members) |member| {
+        const field = container.handle.tree.fullContainerField(member) orelse continue;
+        const name = container.handle.tree.tokenSlice(field.ast.main_token);
+        try completions.append(arena, .{
+            .label = name,
+            .kind = if (field.ast.tuple_like) .EnumMember else .Field,
+            .detail = Analyser.getContainerFieldSignature(container.handle.tree, field),
+            .insertText = name,
+            .insertTextFormat = .PlainText,
+        });
+    }
+}
+
+/// Resolves `identifier`/`path.to.identifier` at `text_index`
+/// If the `identifier` is a container `fn_arg_index` is unused
+/// If the `identifier` is a `fn_name`/`identifier.fn_name`, tries to resolve
+///         `fn_name`'s `fn_arg_index`'s param type
+fn collectContainerNodes(
+    document_store: *DocumentStore,
+    analyser: *Analyser,
+    arena: std.mem.Allocator,
+    handle: *const DocumentStore.Handle,
+    text_index: usize,
```

Comment:
`source index` is the term used to refer to byte indices into document source code throughout the codebase.
```suggestion
    source_index: usize,
```